src/simulations/tests/frequency_set_generator_test.py

```diff
@@ -0,0 +1,195 @@
+      self.assertLen(hists, 3)
+      self.assertEqual(hists[0], [2])
+      self.assertEqual(hists[1], [1,1])
+      self.assertEqual(hists[2], [2,2,2])
+
+  def test_truncated_poisson_pmf(self):
+      h = HomogeneousMultiSetGenerator(10, [1], [1], np.random.RandomState())
+      e = np.exp(1)
+      self.assertEqual(h._truncated_poisson_pmf(1, 1), [1])
+      pmf1 = h._truncated_poisson_pmf(1,4)
+      self.assertLen(pmf1, 4)
+      self.assertAlmostEqual(pmf1[0], 0.3678794)
+      self.assertAlmostEqual(pmf1[1], 0.3678794)
+      self.assertAlmostEqual(pmf1[2], 0.1839397)
+      self.assertAlmostEqual(pmf1[3], 0.0803014)
+      pmf2 = h._truncated_poisson_pmf(2,3)
+      self.assertLen(pmf2, 3)
+      self.assertAlmostEqual(pmf2[0], 0.1353353)
+      self.assertAlmostEqual(pmf2[1], 0.2706706)
+      self.assertAlmostEqual(pmf2[2], 0.5939942)
+  
+  @parameterized.parameters((100, [1, 2], (5,1)), (3, [1,], (1,)))
+  def test_homogeneous_multiset_generator_freq_cap(
+      self, freq_cap, set_sizes, freq_rates):
+    gen = HomogeneousMultiSetGenerator(
+        universe_size=4,
+        set_sizes=set_sizes,
+        freq_rates=freq_rates,
+        freq_cap=freq_cap,
+        random_state=np.random.RandomState(1))
+    output_multiset_ids_list = [multiset_ids for multiset_ids in gen]
+    output_multiset_sizes = [len(set(m)) for m in output_multiset_ids_list]
+    self.assertEqual(output_multiset_sizes, set_sizes)
+
+  def test_homogeneous_multiset_generator_raise_unequal_length_input(self):
+    # Test if raise error when set_sizes and freq_rate_list do not have
+    # equal length.
+    with self.assertRaises(AssertionError):
+      _ = HomogeneousMultiSetGenerator(
+          universe_size=4,
+          set_sizes=[1, 1],
+          freq_rates=[1],
+          freq_cap=3,
+          random_state=np.random.RandomState())
+
+  def test_homogeneous_multiset_generator_raise_invalid_freq_rate(self):
+    # Test if raise error when freq_rate is invalid.
+    with self.assertRaises(AssertionError):
+      _ = HomogeneousMultiSetGenerator(
+          universe_size=4,
+          set_sizes=[1, 1],
+          freq_rates=[-1, 1],
+          freq_cap=3,
+          random_state=np.random.RandomState())
+
+  @parameterized.parameters(0, -1)
+  def test_homogeneous_multiset_generator_raise_invalid_freq_cap(self,
+                                                                 freq_cap):
+    # Test if raise error when freq_cap is invalid.
+    with self.assertRaises(AssertionError):
+      _ = HomogeneousMultiSetGenerator(
+          universe_size=4,
+          set_sizes=[1, 1],
+          freq_rates=[1, 1],
+          freq_cap=freq_cap,
+          random_state=np.random.RandomState())
+
+  def test_homogeneous_multiset_generator_factory_with_num_and_size(self):
+      f = HomogeneousMultiSetGenerator.get_generator_factory_with_num_and_size(
+          100, 3, 5, [1, 2, 3], 10)
+      gen = f(np.random.RandomState(1))
+      sets = [s for s in gen]
```

Comment:
no need to have list comprehension here?

Reply:
Let me know if you like this better.